h/services/list_groups.py

```diff
@@ -67,17 +68,52 @@ def open_groups(self, user=None, authority=None, document_uri=None):
                       .all())
         return self._sort(groups)
 
-    def private_groups(self, user=None):
+    def _private_groups(self, user=None):
         """Return this user's private groups per user.groups."""
 
         if user is None:
             return []
         return self._sort(user.groups)
 
+    def _parse_origin(self, uri):
+        """
+        Return the origin of a URI or None if empty or invalid.
+
+        Per https://tools.ietf.org/html/rfc6454#section-7 :
+        Return ``<scheme> + '://' + <host> + <port>``
+        for a URI.
+
+        :param uri: URI string
+        """
+
+        if uri is None:
+            return None
+        parsed = urlparse.urlsplit(uri)
+        # netloc contains both host and port
+        origin = urlparse.SplitResult(parsed.scheme, parsed.netloc, '', '', '')
+        return origin.geturl() or None
+
     def _sort(self, groups):
         """ sort a list of groups of a single type """
         return sorted(groups, key=lambda group: (group.name.lower(), group.pubid))
 
+    def _world_group(self, authority):
```

Comment:
I'm continuing the existing assumption that the world group _is_ scoped to an authority.